[SPARK-9552] Return "false" while nothing to kill in killExecutors

GraceH opened the pull request:
In discussion (SPARK-9552), we proposed a force kill in `killExecutors`. But if there is nothing to kill, it will return back with true (acknowledgement). And then, it causes the certain executor(s) (which is not eligible to kill) adding to pendingToRemove list for further actions. 

In this patch, we'd like to change the return semantics. If there is nothing to kill, we will return "false". and therefore  all those non-eligible executors won't be added to the pendingToRemove list. 

@vanzin @andrewor14 As the follow up of PR#7888, please let me know your comments.

GraceH committed:
return false is nothing to kill in killExecutors

SparkQA commented:
**[Test build #46186 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46186/consoleFull)** for PR 9796 at commit [`589083b`](https://github.com/apache/spark/commit/589083b8d9208cb1044556d0571f4dfd71284263).

SparkQA commented:
**[Test build #46187 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46187/consoleFull)** for PR 9796 at commit [`0d8e6c1`](https://github.com/apache/spark/commit/0d8e6c11bebb34d0b612c9df5637916806f6c6e2).

andrewor14 commented:
@GraceH please put `[SPARK-9552] in JIRA title

SparkQA commented:
**[Test build #46186 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46186/consoleFull)** for PR 9796 at commit [`589083b`](https://github.com/apache/spark/commit/589083b8d9208cb1044556d0571f4dfd71284263).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46186/
Test PASSed.

SparkQA commented:
**[Test build #46187 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46187/consoleFull)** for PR 9796 at commit [`0d8e6c1`](https://github.com/apache/spark/commit/0d8e6c11bebb34d0b612c9df5637916806f6c6e2).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46187/
Test FAILed.

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -462,7 +463,8 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
    * @param replace whether to replace the killed executors with new ones
    * @param force whether to force kill busy executors
-   * @return whether the kill request is acknowledged.
+   * @return whether the kill request is acknowledged. If list to kill is empty, it should return
```
/should/will

GraceH replied:
thanks a lot. i will change that.

vanzin commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -408,7 +408,8 @@ private[spark] class ExecutorAllocationManager(
     } else {
-      logWarning(s"Unable to reach the cluster manager to kill executor $executorId!")
+      logWarning(s"Unable to reach the cluster manager to kill executor $executorId!" +
+        s"Or no executor to kill!")
```
Two things about this now:

- the message reads a little bit weirdly. How about: "Unable to send kill request for $executorId to cluster manager, or executor is busy."

- warning seems bit strong when the most probable cause of returning false is that the executor is busy; if there's a failure in sending a message to the cluster manager, I'm sure we'll see error / warning logs from other parts of the code.

GraceH replied:
1. or maybe we need to change that to "or executor is not eligible to kill". since there are two situations to get empty list: a) busy b) kill that repeatedly

2. There is not action to take in other parts of the code. The problem here is if we not change the API (return type). it is really hard to tell it is not acknowledged or no executor to kill. If it is not acknowledged, we should print warning, but if no executor to kill, it is the normal case with info level.

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -450,7 +450,8 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
   /**
    * Request that the cluster manager kill the specified executors.
-   * @return whether the kill request is acknowledged.
+   * @return whether the kill request is acknowledged. If list to kill is empty, it should return
```
s/should/will

GraceH replied:
thanks. will change that.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #46224 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46224/consoleFull)** for PR 9796 at commit [`0d8e6c1`](https://github.com/apache/spark/commit/0d8e6c11bebb34d0b612c9df5637916806f6c6e2).

andrewor14 commented on core/src/test/scala/org/apache/spark/deploy/StandaloneDynamicAllocationSuite.scala:
```diff
@@ -395,8 +395,8 @@ class StandaloneDynamicAllocationSuite
-    // kill executor 1
-    assert(sc.killExecutor(executors.head))
+    // kill executor 1, and actually nothing to kill
+    assert(sc.killExecutor(executors.head) === false)
```
just do `!sc.killExecutor`

andrewor14 replied:
why is there actually nothing to kill?

GraceH replied:
because this one is killed in replacement part. 
```java
assert(executors.size === 2)
    // kill executor 1, and replace it
    assert(sc.killAndReplaceExecutor(executors.head)) //executors.head is killed here with replace = true.

GraceH committed:
addressing the comments

SparkQA commented:
**[Test build #46253 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46253/consoleFull)** for PR 9796 at commit [`0d8e6c1`](https://github.com/apache/spark/commit/0d8e6c11bebb34d0b612c9df5637916806f6c6e2).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46253/
Test FAILed.

GraceH committed:
addressing the comments

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46273/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46274/
Test FAILed.

andrewor14 commented:
retest this please

andrewor14 commented on core/src/test/scala/org/apache/spark/deploy/StandaloneDynamicAllocationSuite.scala:
```diff
@@ -395,8 +395,8 @@ class StandaloneDynamicAllocationSuite
-    // kill executor 1
-    assert(sc.killExecutor(executors.head))
+    // kill executor 1, and actually nothing to kill
+    assert(!sc.killExecutor(executors.head))
```
I think this may be a bigger problem. This is supposed to be `true` because it's a new executor so we should be able to kill it. I think it's the way this test is set up that makes this false, because the driver doesn't wait for new executors to come up. We might need to do some more mocking here.

GraceH replied:
@andrewor14  The executors.head is assigned beforehand. for example, you have two executor ID {27,28}. Then, the first one(id 27) is killed with replacement. But I guess the newly created executor cannot be with the same ID. After that, you try to kill the header executor (id 27), it should return empty list (since 27 has been in the pendingToRemove list). Am I right?

andrewor14 replied:
no, the idea is more like the following:

- you start with executors {27, 28}
- you kill and replace 27, so you end up with executors {28, 29}
- now you want to kill 28, this should succeed (but currently it doesn't in the tests)

GraceH replied:
if so, we should not kill excutors.head(27). it should be excutor(1). am I right?

���� Grace from mobile phone

-------- �������� --------
������Re: [spark] [SPARK-9552] Return "false" while nothing to kill in killExecutors (#9796)
��������andrewor14
��������apache/spark
������"Huang, Jie"


In core/src/test/scala/org/apache/spark/deploy/StandaloneDynamicAllocationSuite.scala<https://github.com/apache/spark/pull/9796#discussion_r45414046>:

> @@ -395,8 +395,8 @@ class StandaloneDynamicAllocationSuite
>      }
>
>      var apps = getApplications()
> -    // kill executor 1
> -    assert(sc.killExecutor(executors.head))
> +    // kill executor 1, and actually nothing to kill
> +    assert(!sc.killExecutor(executors.head))


no, the idea is more like the following:

  *   you start with executors {27, 28}
  *   you kill and replace 27, so you end up with executors {28, 29}
  *   now you want to kill 28, this should succeed (but currently it doesn't in the tests)

��
Reply to this email directly or view it on GitHub<https://github.com/apache/spark/pull/9796/files#r45414046>.

GraceH replied:
according to my understanding, the 1st case tries to kill 27. the 2nd one is to kill 28. that is why the first one causes nothing to happen. the latter case actually kills the executor successfully.

btw, we donot change the 'val executors' after the first assignment.

GraceH replied:
@andrewor14 you can find there are two test cases.  I guess the second one is that you want.

andrewor14 replied:
let's move the discussion to the main thread

SparkQA commented:
**[Test build #46340 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46340/consoleFull)** for PR 9796 at commit [`4a6d06e`](https://github.com/apache/spark/commit/4a6d06ee4a93671daf3956f0d4f86c0d45e89958).

SparkQA commented:
**[Test build #46340 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46340/consoleFull)** for PR 9796 at commit [`4a6d06e`](https://github.com/apache/spark/commit/4a6d06ee4a93671daf3956f0d4f86c0d45e89958).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46340/
Test FAILed.

SparkQA commented:
**[Test build #46653 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46653/consoleFull)** for PR 9796 at commit [`657849d`](https://github.com/apache/spark/commit/657849d153922d1c1a2f909b94f257b6450f4ee6).

SparkQA commented:
**[Test build #46653 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46653/consoleFull)** for PR 9796 at commit [`657849d`](https://github.com/apache/spark/commit/657849d153922d1c1a2f909b94f257b6450f4ee6).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46653/
Test FAILed.

andrewor14 commented:
@GraceH continuing our discussion on the "pending replacement" test:

The problem is actually how the framework is set up. To speed up the test we don't wait for new executors to register, so actually this test was incorrectly written even before this patch. This is because:

- First we start with 2 executors, let's say {27, 28}
- Then we kill and replace 1. What we *should* end up with is {28, 29}, but in this test we don't wait for executor 29 to come up, so we still have {27, 28}.
- Then we try to kill 27 again, and this fails and that's why `sc.killExecutor(27)` was false.

The right fix would be to add an `eventually` block in the test to wait until we have 2 executors but with different IDs. We can do this by comparing `executorIdsBefore != executorIdsAfter` before and after the call to `sc.killAndReplace`. Even then some of the following test logic has to change.

Fixing this test may be fairly involved. Will you have time to look into this? If not, I can take over later after the 1.6 release.

GraceH commented:
@andrewor14 Yes. you are so right. Meanwhile it seems the original implementation has waited for a while to check if the replacement is there.  According to you suggestion, I can add the `executor id comparison` here. And it is tested locally. What do you think? 

```Java
val executors = getExecutorIds(sc)
    assert(executors.size === 2)
    // kill executor 1, and replace it
    assert(sc.killAndReplaceExecutor(executors.head))
    eventually(timeout(10.seconds), interval(10.millis)) {
      val apps = getApplications()
      assert(apps.head.executors.size === 2)
+     // make sure the old executors head has been killedAndReplaced
+     assert(executors.head != getExecutorIds(sc).head)
    }

GraceH commented:
I have added the test case https://github.com/GraceH/spark/commit/2e4884c30d9edb0a366e9138cbad8772c5645c5d. Please let me know your comments.

GraceH committed:
check if the old executor head is killed and replaced

andrewor14 commented on core/src/test/scala/org/apache/spark/deploy/StandaloneDynamicAllocationSuite.scala:
```diff
@@ -386,17 +386,21 @@ class StandaloneDynamicAllocationSuite
     eventually(timeout(10.seconds), interval(10.millis)) {
       val apps = getApplications()
       assert(apps.head.executors.size === 2)
+      val executorIdsAfter = getExecutorIds(sc).head
```
why not just compare the `Seq`? E.g. `val executorIdsAfter = getExecutorIds(sc)`

andrewor14 commented on core/src/test/scala/org/apache/spark/deploy/StandaloneDynamicAllocationSuite.scala:
```diff
@@ -386,17 +386,21 @@ class StandaloneDynamicAllocationSuite
+    // kill executor 1, and actually nothing to kill
+    assert(!sc.killExecutor(executors.head))
     apps = getApplications()
     assert(apps.head.executors.size === 2)
```
this may need to change. I think after killing an executor here the size should be 1, not 2.

GraceH replied:
since the first attempt is to kill {27}, nothing happens. The executors.size should be fine with 2.

andrewor14 replied:
what? I thought we want to kill `28` here

GraceH replied:
since we didn't change the `executors` (essentially it is the `executorIdsBefore`)  after the previous assignment. it is still {27}.  `executors(1) == 28`.

GraceH replied:
That is why killing nothing, and executors.size == 2. Since the previous 27 was killed. 
The real kill ({28}) is tested in the following case.

andrewor14 replied:
we should find a way to change the `executors` list. Right now it's totally arbitrary that we kill the second executor but not the first. Anyone else other than us will be confused when he/she reads this code

andrewor14 replied:
in other words, the goal here is to kill `executors.head` but have that be 28, and still have all the asserts passing.

GraceH replied:
how about this?

```java
  val executorKilledAndReplaced = executors.head // The previous head which is killed and replaced.
  assert(!sc.killExecutor(executorKilledAndReplaced))
  val executorToKill = executors(1) // The valid executor which is still working
  assert(sc.killExecutor(executorToKill))
  val executorAppended = executors(2) // The newly added executor (replacement)
  assert(sc.killExecutor(executorAppended))

andrewor14 replied:
That still uses the old `executors` list. All we need to do is refresh it from `sc`: https://github.com/GraceH/spark/pull/2

andrewor14 commented:
Do the tests pass locally?

GraceH commented:
Yes. The replacement is finished.

SparkQA commented:
**[Test build #46726 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46726/consoleFull)** for PR 9796 at commit [`2e4884c`](https://github.com/apache/spark/commit/2e4884c30d9edb0a366e9138cbad8772c5645c5d).

GraceH committed:
to compare the seq instead of head only

SparkQA commented:
**[Test build #46728 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46728/consoleFull)** for PR 9796 at commit [`154ab31`](https://github.com/apache/spark/commit/154ab3156aaacd0ccb026a5ec1d178a25935edea).

SparkQA commented:
**[Test build #46726 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46726/consoleFull)** for PR 9796 at commit [`2e4884c`](https://github.com/apache/spark/commit/2e4884c30d9edb0a366e9138cbad8772c5645c5d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46726/
Test PASSed.

SparkQA commented:
**[Test build #46728 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46728/consoleFull)** for PR 9796 at commit [`154ab31`](https://github.com/apache/spark/commit/154ab3156aaacd0ccb026a5ec1d178a25935edea).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46728/
Test FAILed.

GraceH commented:
retest this please

SparkQA commented:
**[Test build #46752 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46752/consoleFull)** for PR 9796 at commit [`154ab31`](https://github.com/apache/spark/commit/154ab3156aaacd0ccb026a5ec1d178a25935edea).

SparkQA commented:
**[Test build #46752 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46752/consoleFull)** for PR 9796 at commit [`154ab31`](https://github.com/apache/spark/commit/154ab3156aaacd0ccb026a5ec1d178a25935edea).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46752/
Test FAILed.

zsxwing commented:
retest this please

SparkQA commented:
**[Test build #47557 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47557/consoleFull)** for PR 9796 at commit [`154ab31`](https://github.com/apache/spark/commit/154ab3156aaacd0ccb026a5ec1d178a25935edea).

SparkQA commented:
**[Test build #47557 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47557/consoleFull)** for PR 9796 at commit [`154ab31`](https://github.com/apache/spark/commit/154ab3156aaacd0ccb026a5ec1d178a25935edea).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47557/
Test PASSed.

GraceH commented:
Thanks @zsxwing. The patch seems to pass all tests.

andrewor14 commented:
This is not ready for merge yet, please see the unresolved comments thread.

Commit:
Fix kill and replace test

Previously we continued to use the old executors list even after
killing and adding a new executor to replace the old one. This
commit ensures subsequent asserts refresh the list.

andrewor14 commented:
@GraceH The "pending replacement" test is still not correct; we never actually updated the executors list after killing and replacing an executor. The test currently passes but it's very brittle. I've opened https://github.com/GraceH/spark/pull/2 as a pull request against your branch as a suggestion on how to fix it.

GraceH commented:
I leave my thoughts under GraceH#2. Thanks.

Commit:
Merge pull request #2 from andrewor14/pr-9796-suggestion

Fix kill and replace test

SparkQA commented:
**[Test build #47965 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47965/consoleFull)** for PR 9796 at commit [`fd5f435`](https://github.com/apache/spark/commit/fd5f4351d347d6b68c39ff19c26cd15c899048b1).

SparkQA commented:
**[Test build #47965 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47965/consoleFull)** for PR 9796 at commit [`fd5f435`](https://github.com/apache/spark/commit/fd5f4351d347d6b68c39ff19c26cd15c899048b1).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47965/
Test FAILed.

Commit:
Add kill test cases for killAndReplace.

1. add test case to check if it is false when killing non-existing executor(killed and replaced)
2. add test case to check if we can kill newly appended executor

SparkQA commented:
**[Test build #47983 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47983/consoleFull)** for PR 9796 at commit [`bf2edd3`](https://github.com/apache/spark/commit/bf2edd34d9c97040f55a10550528fa7aefd7417a).

SparkQA commented:
**[Test build #47983 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47983/consoleFull)** for PR 9796 at commit [`bf2edd3`](https://github.com/apache/spark/commit/bf2edd34d9c97040f55a10550528fa7aefd7417a).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47983/
Test FAILed.

Commit:
eliminate space in blank line

SparkQA commented:
**[Test build #47984 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47984/consoleFull)** for PR 9796 at commit [`0305815`](https://github.com/apache/spark/commit/030581501f54f4d1052f673f2e4a25f94fb61455).

SparkQA commented:
**[Test build #47984 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47984/consoleFull)** for PR 9796 at commit [`0305815`](https://github.com/apache/spark/commit/030581501f54f4d1052f673f2e4a25f94fb61455).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47984/
Test PASSed.

andrewor14 commented on core/src/test/scala/org/apache/spark/deploy/StandaloneDynamicAllocationSuite.scala:
```diff
@@ -386,23 +386,28 @@ class StandaloneDynamicAllocationSuite
+    syncExecutors(sc)
+
+    // kill newly created executor and do not replace it
+    assert(sc.killExecutor(newExecutors(1)))
```
why not just kill `newExecutors.head`? This is kind of arbitrary

andrewor14 commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -408,7 +408,8 @@ private[spark] class ExecutorAllocationManager(
       true
     } else {
-      logWarning(s"Unable to reach the cluster manager to kill executor $executorId!")
+      logWarning(s"Unable to reach the cluster manager to kill executor $executorId," +
```
need space after comma

andrewor14 replied:
I'll fix this myself

andrewor14 commented:
LGTM thanks @GraceH and @vanzin I'm merging this into master.

GraceH commented:
@andreor14 thanks.
